lib/components/metrics.js

```diff
@@ -0,0 +1,293 @@
+
+/**
+ * Prometheus-style /metrics gathering and exporting.
+ * This class provides a central location to register gauge and counter metrics
+ * used to generate the <code>/metrics</code> page. It also contains collector
+ * code to generate the standard <code>process_</code> and <code>nodejs_</code>
+ * metrics:
+ *   process_resident_memory_bytes
+ *   process_virtual_memory_bytes
+ *   process_heap_bytes
+ *   process_cpu_seconds_total
+ *   process_open_fds
+ *   process_max_fds
+ *   process_start_time_seconds
+ *   nodejs_heap_used_bytes
+ *
+ * @example <caption>A simple metric that counts the keys in an object:</caption>
+ *   var metrics = new Metrics();
+ *
+ *   var orange = {};
+ *   metrics.addGauge({
+ *       name: "oranges",
+ *       help: "current number of oranges",
+ *       refresh: (gauge) => {
+ *           gauge.set({}, Object.keys(oranges).length);
+ *       },
+ *   });
+ *
+ * @example <caption>Generating values for multiple gauges in a single collector function.</caption>
+ *   var metrics = new Metrics();
+ *
+ *   var oranges_gauge = metrics.addGauge({
+ *       name: "oranges",
+ *       help: "current number of oranges",
+ *   });
+ *   var apples_gauge = metrics.addGauge({
+ *       name: "apples",
+ *       help: "current number of apples",
+ *   });
+ *
+ *   metrics.addCollector(() => {
+ *       var counts = this._countFruit();
+ *       oranges_gauge.set({}, counts.oranges);
+ *       apples_gauge.set({}, counts.apples);
+ *   });
+ *
+ * @example <caption>Using counters</caption>
+ *   var metrics = new Metrics();
+ *
+ *   metrics.addCollector({
+ *       name: "things_made",
+ *       help: "count of things that we have made",
+ *   });
+ *
+ *   function makeThing() {
+ *       metrics.incCounter("things_made");
+ *       return new Thing();
+ *   }
+ *
+ * @constructor
+ */
+function Metrics() {
+    // Only attempt to load these dependencies if metrics are enabled
+    var Prometheus = require("prometheus-client");
+
+    var client = this._client = new Prometheus();
+
+    this._collectors = []; // executed in order
+    this._counters = {};
+
+    // Register some built-in process-wide metrics
+    // See also
+    //   https://prometheus.io/docs/instrumenting/writing_clientlibs/#standard-and-runtime-collectors
+
+    var rss_gauge = this.addGauge({
+        namespace: "process",
+        name: "resident_memory_bytes",
+        help: "Resident memory size in bytes",
+    });
+    var vsz_gauge = this.addGauge({
+        namespace: "process",
+        name: "virtual_memory_bytes",
+        help: "Virtual memory size in bytes",
+    });
+
+    var heap_size_gauge = this.addGauge({
+        namespace: "process",
+        name: "heap_bytes",
+        help: "Total size of Node.js heap in bytes",
+    });
+    var heap_used_gauge = this.addGauge({
+        namespace: "nodejs",
+        name: "heap_used_bytes",
+        help: "Used size of Node.js heap in bytes",
```

Comment:
This is all wonderful :heart: 

Why don't we [do something similar](https://github.com/matrix-org/matrix-appservice-bridge/pull/30#discussion_r83233542) with `AdminCommands`? :angel: